File: python/examples/benchmark.py
#!/usr/bin/env python3
"""Benchmark for MagicQueue Python SDK."""

import time
import threading
from magicqueue import Queue, Worker, Job, JobContext

TOTAL_JOBS = 10000
BATCH_SIZE = 1000
WORKER_BATCH_SIZE = 50

completed = 0
lock = threading.Lock()


def handler(job: Job, ctx: JobContext):
    global completed
    # Minimal work
    _ = job.data
    with lock:
        completed += 1


def main():
    global completed

    print("=== MagicQueue Python Benchmark ===\n")
    print(f"Jobs: {TOTAL_JOBS:,}")
    print(f"Push batch size: {BATCH_SIZE:,}")
    print(f"Worker batch size: {WORKER_BATCH_SIZE}")
    print()

    # Start worker in background thread
    worker = Worker("benchmark", handler, batch_size=WORKER_BATCH_SIZE)
    worker_thread = threading.Thread(target=worker.start, daemon=True)
    worker_thread.start()
    time.sleep(0.3)

    # Batch push
    print("--- Batch Push ---")
    queue = Queue("benchmark")
    queue.connect()

    push_start = time.time()

    for batch_start in range(0, TOTAL_JOBS, BATCH_SIZE):
        batch_end = min(batch_start + BATCH_SIZE, TOTAL_JOBS)
        jobs = [{"data": {"i": i}} for i in range(batch_start, batch_end)]
        queue.push_batch(jobs)

    push_duration = time.time() - push_start
    push_throughput = TOTAL_JOBS / push_duration

    print(f"Push duration: {push_duration:.2f}s")
    print(f"Push throughput: {push_throughput:,.0f} jobs/sec")
    print()

    queue.close()

    # Wait for processing
    print("--- Processing ---")
    process_start = time.time()

    while completed < TOTAL_JOBS:
        pct = (completed / TOTAL_JOBS) * 100
        elapsed = time.time() - process_start
        rate = completed / elapsed if elapsed > 0 else 0
        print(f"\rProcessed: {completed:,} ({pct:.1f}%) - {rate:,.0f}/s", end="", flush=True)
        time.sleep(0.5)

    process_duration = time.time() - process_start
    process_throughput = TOTAL_JOBS / process_duration

    print(f"\rProcess duration: {process_duration:.2f}s                    ")
    print(f"Process throughput: {process_throughput:,.0f} jobs/sec")
    print()

    # Total
    total_duration = time.time() - push_start
    print("=== RESULTS ===")
    print(f"Total: {total_duration:.2f}s")
    print(f"Throughput: {TOTAL_JOBS / total_duration:,.0f} jobs/sec")

    worker.stop()


if __name__ == "__main__":
    main()

File: client/examples/benchmark.ts
import { Queue, Worker, Job } from "../src";

const TOTAL_JOBS = 100_000;
const BATCH_SIZE = 5000;
const WORKER_BATCH_SIZE = 50;

async function benchmark() {
  console.log("=== MagicQueue ULTRA Benchmark ===\n");
  console.log(`Jobs: ${TOTAL_JOBS.toLocaleString()}`);
  console.log(`Push batch size: ${BATCH_SIZE.toLocaleString()}`);
  console.log(`Worker batch size: ${WORKER_BATCH_SIZE}\n`);

  let completed = 0;

  // Worker con batch processing
  const worker = new Worker(
    "benchmark",
    async (_job: Job) => {
      completed++;
    },
    { batchSize: WORKER_BATCH_SIZE }
  );

  const workerPromise = worker.start();
  await new Promise((r) => setTimeout(r, 300));

  // === BATCH PUSH ===
  const queue = new Queue("benchmark");
  await queue.connect();

  console.log("--- Batch Push ---");
  const pushStart = performance.now();

  const batches = Math.ceil(TOTAL_JOBS / BATCH_SIZE);
  for (let b = 0; b < batches; b++) {
    const batchJobs = [];
    const start = b * BATCH_SIZE;
    const end = Math.min(start + BATCH_SIZE, TOTAL_JOBS);

    for (let i = start; i < end; i++) {
      batchJobs.push({ data: { i }, priority: 0 });
    }

    await queue.pushBatch(batchJobs);

    if ((b + 1) % 10 === 0) {
      process.stdout.write(`\rPushed: ${((b + 1) * BATCH_SIZE).toLocaleString()}`);
    }
  }

  const pushEnd = performance.now();
  const pushDuration = (pushEnd - pushStart) / 1000;
  const pushThroughput = TOTAL_JOBS / pushDuration;

  console.log(`\rPush duration: ${pushDuration.toFixed(2)}s                    `);
  console.log(`Push throughput: ${pushThroughput.toLocaleString(undefined, { maximumFractionDigits: 0 })} jobs/sec\n`);

  await queue.close();

  // === PROCESSING ===
  console.log("--- Processing (batch pull + batch ack) ---");
  const processingStart = performance.now();

  const progressInterval = setInterval(() => {
    const pct = ((completed / TOTAL_JOBS) * 100).toFixed(1);
    const elapsed = (performance.now() - processingStart) / 1000;
    const rate = completed / elapsed;
    process.stdout.write(`\rProcessed: ${completed.toLocaleString()} (${pct}%) - ${rate.toLocaleString(undefined, { maximumFractionDigits: 0 })}/s`);
  }, 500);

  while (completed < TOTAL_JOBS) {
    await new Promise((r) => setTimeout(r, 100));
  }

  clearInterval(progressInterval);

  const processEnd = performance.now();
  const processDuration = (processEnd - processingStart) / 1000;
  const processThroughput = TOTAL_JOBS / processDuration;

  console.log(`\rProcess duration: ${processDuration.toFixed(2)}s                                        `);
  console.log(`Process throughput: ${processThroughput.toLocaleString(undefined, { maximumFractionDigits: 0 })} jobs/sec\n`);

  // === TOTALE ===
  const totalDuration = (processEnd - pushStart) / 1000;
  console.log("=== RESULTS ===");
  console.log(`Total: ${totalDuration.toFixed(2)}s`);
  console.log(`Throughput: ${(TOTAL_JOBS / totalDuration).toLocaleString(undefined, { maximumFractionDigits: 0 })} jobs/sec`);

  await worker.stop();
  process.exit(0);
}

benchmark().catch(console.error);
